pkg/issuer/acme/http/istio.go

```diff
@@ -0,0 +1,201 @@
+	return nil
+}
+
+func (s *Solver) getVirtualService(ctx context.Context, ch *cmacme.Challenge) (*unstructured.Unstructured, error) {
+	selector := labels.Set(podLabels(ch)).AsSelector()
+	vsList, err := s.virtualServiceLister.Namespace(ch.Namespace).List(selector)
+	if err != nil {
+		return nil, err
+	}
+	switch len(vsList) {
+	case 0:
+		return nil, nil
+	case 1:
+		return vsList[0], nil
+	default:
+		for _, vs := range vsList[1:] {
+			s.DynamicClient.Resource(util.VirtualServiceGvr()).Namespace(ch.Namespace).Delete(ctx, vs.GetName(), metav1.DeleteOptions{})
+		}
+		return nil, fmt.Errorf("multiple VirtualServices found")
+	}
+}
+
+func (s *Solver) createVirtualService(ctx context.Context, ch *cmacme.Challenge, svcName string) (*unstructured.Unstructured, error) {
+	expectedSpec := createVirtualServiceSpec(ch, svcName)
+
+	vs := newUnstructuredVirtualService(
+		util.VirtualSerivceGvk().Group,
+		util.VirtualSerivceGvk().Version,
+		util.VirtualSerivceGvk().Kind,
+		"cm-acme-http-solver-",
+		ch.Namespace,
+		podLabels(ch),
+		&[]metav1.OwnerReference{*metav1.NewControllerRef(ch, challengeGvk)},
+		expectedSpec,
+	)
+
+	val, err := s.DynamicClient.Resource(util.VirtualServiceGvr()).Namespace(ch.Namespace).Create(ctx, vs, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return val, nil
+}
+
+func (s *Solver) checkAndUpdateVirtualService(ctx context.Context, ch *cmacme.Challenge, svcName string, virtualservice *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	log := logf.FromContext(ctx, "checkAndUpdateVirtualService")
+
+	expectedSpec := createVirtualServiceSpec(ch, svcName)
+
+	spec, err := unstructuredGetSpec(virtualservice)
+	if err != nil {
+		return nil, err
+	}
+	if reflect.DeepEqual(spec, expectedSpec) {
+		return virtualservice, nil
+	}
+
+	log.Info("Updating VirtualService")
+
+	if err := unstructuredSetSpec(virtualservice, expectedSpec); err != nil {
+		return nil, err
+	}
+	val, err := s.DynamicClient.Resource(util.VirtualServiceGvr()).Namespace(ch.Namespace).Update(ctx, virtualservice, metav1.UpdateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return val, nil
+}
+
+func createVirtualServiceSpec(ch *cmacme.Challenge, svcName string) *map[string]interface{} {
+	http01Istio := ch.Spec.Solver.HTTP01.Istio
+	return &map[string]interface{}{
+		"exportTo": []interface{}{"*"},
+		"hosts":    []interface{}{ch.Spec.DNSName},
+		"gateways": []interface{}{http01Istio.GatewayNamespace + "/" + http01Istio.GatewayName},
+		"http": []interface{}{
+			map[string]interface{}{
+				"match": []interface{}{
+					map[string]interface{}{
+						"uri": map[string]interface{}{
+							"exact": solverPathFn(ch.Spec.Token),
+						},
+					},
+				},
+				"route": []interface{}{
+					map[string]interface{}{
+						"destination": map[string]interface{}{
+							"host": fmt.Sprintf("%s.%s.svc.cluster.local", svcName, ch.Namespace),
```

Comment:
`cluster.local` is a globally configurable option on Kubernetes. Is there a way we can determine the domain of the cluster we are running in?
 
Is it sufficient to just do `<service-name>.<namespace>.svc`?